integrations/tensorflow/e2e/math/math_test.py

```diff
@@ -0,0 +1,561 @@
+  wrapped_function = lambda *args: tf_math_function(*args, **config.kwargs)
+
+  if FLAGS.dynamic_dims:
+    signature = tf_utils.apply_function(signature, _make_dims_dynamic)
+
+  return tf_test_utils.tf_function_unittest(
+      input_signature=signature,
+      input_generator=config.generator,
+      input_args=config.args,
+      name=exported_name)(wrapped_function)
+
+
+def create_unittests(module: Union["TfMathModule", Type["TfMathModule"]]):
+  """Creates all tf_function_unittests to test and puts them on module."""
+  for function_name in FLAGS.functions:
+    # pytype: disable=attribute-error :\
+    for exported_name, config in FUNCTION_TO_CONFIGS[function_name].items():
+      # pytype: enable=attribute-error :/
+      if is_complex(config) and not FLAGS.test_complex:
+        continue
+      function_unittest = create_function_unittest(config, function_name,
+                                                   exported_name)
+      setattr(module, exported_name, function_unittest)
+
+
+class TfMathModule(tf_test_utils.TestModule):
+
+  @classmethod
+  def configure_class(cls):
+    create_unittests(cls)
+
+  def __init__(self):
+    super().__init__()
+    create_unittests(self)
+
+
+class TfMathTest(tf_test_utils.TracedModuleTestCase):
+
+  def __init__(self, *args, **kwargs):
+    super().__init__(*args, **kwargs)
+    self._modules = tf_test_utils.compile_tf_module(
+        TfMathModule, exported_names=TfMathModule.get_exported_names())
+
+
+def main(argv):
+  del argv  # Unused.
+  if hasattr(tf, "enable_v2_behavior"):
+    tf.enable_v2_behavior()
+
+  if FLAGS.list_functions_with_complex_tests:
+    for function_name, configs in FUNCTION_TO_CONFIGS.items():
+      # pytype: disable=attribute-error :\
+      for exported_name, config in configs.items():
+        # pytype: enable=attribute-error :/
+        if is_complex(config):
+          print(f'    "{function_name}",')
+    return
+
+  if len(FLAGS.functions) > 1:
+    # We only allow testing multiple functions with a single target backend
+    # so that we can store the artifacts under:
+    #   'artifacts_dir/multiple_functions__backend/...'
+    # We specialize the 'multiple_functions' dir by backend to avoid overwriting
+    # tf_input.mlir and iree_input.mlir. These are typically identical across
+    # backends, but are not when the functions to compile change per-backend.
+    if len(FLAGS.target_backends) != 1:
+      raise flags.IllegalFlagValueError(
+          "Expected len(target_backends) == 1 when len(functions) > 1, but got "
+          f"the following values for target_backends: {FLAGS.target_backends}.")
+    function_str = f"multiple_functions__{FLAGS.target_backends[0]}"
+  else:
+    function_str = FLAGS.functions[0]
+  dim_str = "dynamic_dims" if FLAGS.dynamic_dims else "static_dims"
+  settings_str = os.path.join(function_str, dim_str)
+  TfMathModule.__name__ = os.path.join("tf", "math", settings_str)
```

Comment:
This seems a little hacky (but ok). Please comment it.

Reply:
Which part, the assignment to `__name__`? We've been doing this for a while (e.g. in the keras vision tests), but it would probably be better to (optionally) pipe in the relative artifacts dir along with the module itself in the test case `__init__`, particularly since we're using multi-purpose modules more and more now.

Reply:
I was mostly referring to effectively using a file path as part of `__name__`

Reply:
Added note.